### PR TITLE
Added --dry-run flag to wp role reset

### PR DESCRIPTION
**Pull request by rahulsprajapati**
Issue : "Add --dry-run flag to wp role reset" #3114

### CODE REVIEW

File: php/commands/role.php [hunk 1]
```diff
@@ -367,6 +383,167 @@ private static function persistence_check() {
 		if ( !$wp_roles->use_db )
 			WP_CLI::error( "Role definitions are not persistent." );
 	}
+
+	private static function get_role_capability( $role_key ) {
```
Comment (danielbachhuber) [inline php/commands/role.php]: Why did you need to create this method? Seems like a lot of code to internalize — not sure if it's a great idea.

Comment (rahulsprajapati) [inline php/commands/role.php]: For comparing the default capability and exist capability of a role in the database, we needed list/array of default capability of a role for dry-run. As for reset role we first remove role and store the capability in the array, then we add the roles with default capability again. After that, we check count using that array.
Here I didn't found any WordPress core function to fetch the default capability of the role. We can add this function in core.
There is one another way without using this method, we can add the removed capability after resetting role for dry-run. But I was not sure if that's a good idea.

Comment (danielbachhuber) [inline php/commands/role.php]: > Here I didn't found any WordPress core function to fetch the default capability of the role.

Oh, sad :(

> There is one another way without using this method, we can add the removed capability after resetting role for dry-run. But I was not sure if that's a good idea because I think in dry-run we should not do any changes in the database.

I agree.

I'm going to focus on getting v0.24.0 out the door for the next couple of days, then I'll spend a bit of time thinking about how we can get around this issue.

Comment (rahulsprajapati) [inline php/commands/role.php]: Yeah, sure. 
Meanwhile, I will try to find the alternative solution. Thank you.

File: php/commands/role.php [hunk 2]
```diff
@@ -258,6 +262,7 @@ public function reset( $args, $assoc_args ) {
 		global $wp_roles;
 		$all_roles = array_keys( $wp_roles->roles );
 		$preserve_args = $args;
+		$this->dry_run         = \WP_CLI\Utils\get_flag_value( $assoc_args, 'dry-run' );
```
Comment (danielbachhuber) [inline php/commands/role.php]: Does this need to be a class variable? And, to avoid negative conditionals throughout, could we name it something like `$alter_database`?

File: php/commands/role.php (hunk 1 above)
Comment (danielbachhuber) [inline php/commands/role.php]: @rahulsprajapati I'm still hesitant about internalizing this array of roles and capabilities because I'm concerned it will likely get out of date.

Have you thought of any other ways to produce the same effect? If not, I think we might want to punt on this.

File: php/commands/role.php (hunk 2 above)
Comment (rahulsprajapati) [inline php/commands/role.php]: No, we don't need dry-run as class variable as we are not using it in any other method. :+1:

File: php/commands/role.php [hunk 3]
```diff
@@ -256,8 +260,13 @@ public function reset( $args, $assoc_args ) {
...
 		global $wp_roles;
-		$all_roles = array_keys( $wp_roles->roles );
-		$preserve_args = $args;
+		$all_roles      = array_keys( $wp_roles->roles );
+		$preserve_args  = $args;
+		$dry_run        = \WP_CLI\Utils\get_flag_value( $assoc_args, 'dry-run' );
+		$alter_database = false;
+		if ( ! $dry_run ) {
+			$alter_database = true;
+		}
```
Comment (rahulsprajapati) [inline php/commands/role.php]: @danielbachhuber Is this looks good ? :)

File: php/commands/role.php (hunk 1 above)
Comment (rahulsprajapati) [inline php/commands/role.php]: @danielbachhuber I was thinking if we can use this array of roles function in WordPress core 

i.e. : https://core.trac.wordpress.org/browser/trunk/src/wp-admin/includes/schema.php#L631

```
function populate_roles_160() {

	// Dummy gettext calls to get strings in the catalog.
	/* translators: user role */
	_x('Administrator', 'User role');
	/* translators: user role */
	_x('Editor', 'User role');
	/* translators: user role */
	_x('Author', 'User role');
	/* translators: user role */
	_x('Contributor', 'User role');
	/* translators: user role */
	_x('Subscriber', 'User role');

	$roles = array(
		'administrator' => 'Administrator',
		'editor'        => 'Editor',
		'author'        => 'Author',
		'contributor'   => 'Contributor',
		'subscriber'    => 'Subscriber'
	);
	foreach ( $roles as $role_key => $role_name ) {
		add_role( $role_key, $role_name );
		$role = get_role( $role_key );
		// We can add second parameter to fetch roles for respective versions
		$role_capabilities = get_role_capability( $role_key, "roles_160" );
		foreach ( $role_capabilities as $role_capability ) {
			$role->add_cap( $role_capability );
		}
	}
}
```
And we can use this in wp-cli, also in future for new user role we need to change/update `get_role_capability` function only. 
Let me know your thought. :)

Comment (danielbachhuber) [inline php/commands/role.php]: > Let me know your thought. :)

I think we should just punt on this enhancement for now. Would you be opposed?

Comment (rahulsprajapati) [inline php/commands/role.php]: OK. No prob. :)
